cpp/src/cylon/table.cpp

```diff
@@ -427,16 +497,298 @@ Status Sort(const std::shared_ptr<Table> &table, const std::vector<int32_t> &sor
+
+  std::vector<std::unique_ptr<cylon::ArrayIndexComparator>> comparators(sort_columns.size());
+  for(int64_t i = 0; i < sort_columns.size(); i++) {
+    std::unique_ptr<cylon::ArrayIndexComparator> comp;
+    CreateArrayIndexComparator(
+      cylon::util::GetChunkOrEmptyArray(concatenated->get_table()->column(sort_columns[i]), 0), &comp, sort_orders[i]
+    );
+    comparators[i] = std::move(comp);
+  }
+
+  auto comp = [&](int a, int b) { // a and b are index of table in `tables`
+    int64_t a_idx = table_indices[a], b_idx = table_indices[b];
+    bool temp = CompareRows(comparators, a, b) > 0;
+    return CompareRows(comparators, a_idx, b_idx) > 0;
+  };
+
+  std::priority_queue<int, std::vector<int>, decltype(comp)> pq(comp);
+
+  for(int i = 0; i < tables.size(); i++) {
+    if(table_indices[i] < table_end_indices[i]) {
+      pq.push(i);
+    }
+  }
+
+  auto ctx = concatenated->GetContext();
+  auto pool = cylon::ToArrowPool(ctx);
+  arrow::Int64Builder filter(pool);
+  RETURN_CYLON_STATUS_IF_ARROW_FAILED(filter.Reserve(concatenated->Rows()));
+
+  std::vector<int> temp_v;
+
+  while(!pq.empty()) {
+    int t = pq.top();
+    pq.pop();
+    // std::cout<<table_indices[t]<<std::endl;
+    temp_v.push_back(table_indices[t]);
+    filter.UnsafeAppend(table_indices[t]);
+    table_indices[t] += 1;
+    if(table_indices[t] < table_end_indices[t]) {
+      pq.push(t);
+    }
+  }
+  
+  CYLON_ASSIGN_OR_RAISE(auto take_arr, filter.Finish());
+  CYLON_ASSIGN_OR_RAISE(auto take_res, (arrow::compute::Take(concatenated->get_table(), take_arr)));
+
+  out = std::make_shared<Table>(ctx, take_res.table());
+
+  return Status::OK();
+}
+
+Status DetermineSplitPoints(const std::vector<std::shared_ptr<Table>>& gathered_tables_include_root,
+                            const std::vector<int>& sort_columns,
+                            const std::vector<bool>& sort_orders,
+                            std::shared_ptr<Table>& split_points, 
+                            const std::shared_ptr<CylonContext>& ctx) {
+  std::shared_ptr<Table> merged_table;
+  RETURN_CYLON_STATUS_IF_FAILED(
+    MergeSortedTable(gathered_tables_include_root, sort_columns, sort_orders, merged_table));
+
+  int num_split_points = std::min(merged_table->Rows(), (int64_t)ctx->GetWorldSize() - 1);
+
+  RETURN_CYLON_STATUS_IF_FAILED(SampleTableUniform(merged_table, num_split_points, split_points, ctx));
+
+  return Status::OK();
+}
+
+Status GetSplitPoints(std::shared_ptr<Table>& sample_result,
+                      const std::vector<int>& sort_columns,
+                      const std::vector<bool>& sort_orders,
+                      int num_split_points,
+                      std::shared_ptr<Table>& split_points) {
+  auto ctx = sample_result->GetContext();
+
+  std::vector<std::shared_ptr<cylon::Table>> gather_results;
+  // net::MPICommunicator comm;
+  RETURN_CYLON_STATUS_IF_FAILED(ctx->GetCommunicator()->AllGather(sample_result, &gather_results));
```

Comment:
i think, we need to gather all sample tables to the first worker. We don't need allgather here.